File: py/Structs.py
# Can't import from Constants due to circular import
TILE_WIDTH = 16
TILE_HEIGHT = 24

class Position:
    def __init__(self, x: int, y: int, z: int) -> None:
        self.x: int = x
        self.y: int = y
        self.z: int = z

    def __repr__(self) -> str:
        return f"({self.x // TILE_WIDTH}, {self.y // TILE_WIDTH}, {self.z // TILE_HEIGHT})"

    def isValid(self) -> bool:
        return (self.x >= 0 and self.y >= 0 and self.z >= 0 and not (self.x == 0 and self.y == 0 and self.z == 0))

class Unit:
    def __init__(self, race: str, diameter: int, standHeight: int, kneelHeight: int) -> None:
        self.race: str = race
        self.diameter: int = diameter
        self.standHalfHeight: float = standHeight / 2
        self.kneelHalfHeight: float = kneelHeight / 2
        self.kneeling: bool = False

    def GetHalfHeight(self, kneeling: bool) -> float:
        return self.kneelHalfHeight if kneeling else self.standHalfHeight

File: cpp/src/Structs.cpp
#include "Structs.h"
#include <ostream>
#include <string>

namespace OpenXcom
{
Position::Position(int xVoxel, int yVoxel, int zVoxel)
    : x{xVoxel}
    , y{yVoxel}
    , z{zVoxel}
{
}

bool Position::isValid() const
{
    return (x >= 0 && y >= 0 && z >= 0 && !(x == 0 && y == 0 && z == 0));
}

Unit::Unit(std::string alienRace, int alienDiameter, int standHeight, int kneelHeight)
    : race{alienRace}
    , diameter{alienDiameter}
    , standHalfHeight{standHeight / 2.0}
    , kneelHalfHeight{(kneelHeight == -1) ? standHeight / 2.0 : kneelHeight / 2.0}
{
}
double Unit::GetHalfHeight(bool kneeling) const
{
    return kneeling ? kneelHalfHeight : standHalfHeight;
}
} // namespace OpenXcom

namespace AccuracyTest
{
DevVals::DevVals(int dev, int numberHits, int numberShots)
    : deviation{dev}
    , numHits{numberHits}
    , numShots{numberShots}
{
}
} // namespace AccuracyTest
